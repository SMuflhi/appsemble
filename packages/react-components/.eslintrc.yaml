overrides:
  - files:
      - 'src/**'
    extends:
      - remcohaszing/react
      - remcohaszing/web
    rules:
      node/no-unpublished-import: off
      react/prop-types: off
      react/require-default-props: off
