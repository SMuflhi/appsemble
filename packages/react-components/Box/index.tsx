import classNames from 'classnames';
import { type ComponentPropsWithoutRef, type ReactNode } from 'react';

/**
 * Render a simple Bulma box component.
 */
export function Box({ children, className, ...props }: ComponentPropsWithoutRef<'div'>): ReactNode {
  return (
    <div className={classNames('box', className)} {...props}>
      {children}
    </div>
  );
}
