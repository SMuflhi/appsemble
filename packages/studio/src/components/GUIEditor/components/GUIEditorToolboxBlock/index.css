.blockFrame {
  display: flex;
  float: left;
  margin-bottom: 18px;
  margin-left: 18px;

  &.selected {
    background: hsl(204, 86%, 53%);
  }

  &:hover {
    cursor: pointer;
  }
}
