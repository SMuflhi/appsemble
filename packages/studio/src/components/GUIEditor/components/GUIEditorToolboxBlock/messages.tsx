import { defineMessages } from 'react-intl';

export const messages = defineMessages({
  blockLogo: '{name} Logo',
  search: 'Search Blocks',
  enterBlockName: 'Enter a block name',
});
