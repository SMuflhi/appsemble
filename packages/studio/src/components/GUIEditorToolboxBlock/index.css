.blockFrame {
  align-items: center;
  background: hsl(0, 0%, 96%);
  display: flex;
  flex-direction: column;
  float: left;
  margin-bottom: 12px;
  margin-left: 12px;
  overflow: hidden !important;
  text-overflow: ellipsis;
  width: 120px;
}

.blockFrame:hover {
  cursor: pointer;
}

.blockFrameSelected {
  align-items: center;
  background: hsl(204, 86%, 53%);
  display: flex;
  flex-direction: column;
  float: left;
  margin-bottom: 12px;
  margin-left: 12px;
  overflow: hidden !important;
  text-overflow: ellipsis;
  width: 120px;
}

.subtext {
  display: inline-block;
  overflow: hidden !important;
  text-align: center;
  text-overflow: ellipsis;
  white-space: nowrap;
  width: 120px;
}

.main {
  flex: 1;
}
