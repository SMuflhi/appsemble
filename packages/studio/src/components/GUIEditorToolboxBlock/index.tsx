import { Icon, Loader } from '@appsemble/react-components';
import { IconName } from '@fortawesome/fontawesome-common-types';
import axios from 'axios';
import React from 'react';

import styles from './index.css';

export interface Block {
  id: string;
  description?: string;
  iconName: IconName;
}

export default function GUIEditorToolboxBlock(params: any): React.ReactElement {
  const [blocks, setBlocks] = React.useState<Block[]>();
  const [selectedBlock, setSelectedBlock] = React.useState<Block>();

  React.useEffect(() => {
    const getBlocks = async (): Promise<void> => {
      setBlocks(undefined);
      const { data } = await axios.get('/api/blocks');
      const blocksArr: Block[] = [];
      data.map((block: Block): Block[] => {
        let iconName: IconName = 'question';
        switch (block.id.split('/')[1]) {
          case 'map':
            iconName = 'map-marked-alt';
            break;
          case 'navigation':
            iconName = 'route';
            break;
          case 'markdown':
            iconName = 'font';
            break;
          case 'stats':
            iconName = 'chart-bar';
            break;
          case 'table':
            iconName = 'table';
            break;
          case 'action-button':
            iconName = 'plus-square';
            break;
          case 'data-loader':
            iconName = 'database';
            break;
          case 'detail-viewer':
            iconName = 'eye';
            break;
          case 'feed':
            iconName = 'pager';
            break;
          case 'filter':
            iconName = 'filter';
            break;
          case 'form':
            iconName = 'poll-h';
            break;
          case 'list':
            iconName = 'list';
            break;
          default:
            iconName = 'question';
            break;
        }

        blocksArr.push({
          id: block.id,
          description: block.description,
          iconName,
        });

        return blocksArr;
      });
      setBlocks(blocksArr);
    };
    getBlocks();
  }, []);

  const onKeyDown = (event: React.KeyboardEvent): void => {
    if (event.key === 'Escape') {
      setSelectedBlock(undefined);
      params.selectedBlock(undefined);
    }
  };

  if (blocks === [] || blocks === undefined) {
    return <Loader />;
  }

  return (
    <div className={styles.main}>
      {blocks.map((block: Block) => (
        <div
          key={block.id}
          className={selectedBlock === block ? styles.blockFrameSelected : styles.blockFrame}
          onClick={() => [setSelectedBlock(block), params.selectedBlock(block)]}
          onKeyDown={() => onKeyDown}
          role="button"
          tabIndex={0}
        >
          <Icon icon={block.iconName} size="large" />
          <h2 className={styles.subtext}>{block.id.split('/')[1]}</h2>
        </div>
      ))}
    </div>
  );
}
