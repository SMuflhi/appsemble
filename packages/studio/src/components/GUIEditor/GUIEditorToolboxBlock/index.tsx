import { InputField, ValuePickerProvider } from '@appsemble/react-components';
import { BlockManifest } from '@appsemble/types';
import { NamedEvent } from '@appsemble/web-utils';
import { ChangeEvent, ReactElement, useCallback, useState } from 'react';
import { FormattedMessage, useIntl } from 'react-intl';

import { GUIEditorBlockItem } from '../GUIEditorBlockItem';
import { messages } from './messages';

interface GUIEditorToolboxBlockProps {
  blocks: BlockManifest[];
  name: string;
  onChange: (event: ChangeEvent<HTMLInputElement>, block: BlockManifest) => void;
  value: BlockManifest;
}

export function GUIEditorToolboxBlock({
  blocks,
  name,
  onChange,
  value,
}: GUIEditorToolboxBlockProps): ReactElement {
  const [searchValue, setSearchValue] = useState<string>('');
  const [filterBlocks, setFilterBlocks] = useState<BlockManifest[]>(blocks);

  const intl = useIntl();

  const onChangeSearch = useCallback(
    (event: NamedEvent, query: string) => {
      setSearchValue(query);
      const updatedList = blocks.filter(
        (item) => item.name.toLowerCase().search(query.toLowerCase()) !== -1,
      );

      setFilterBlocks(updatedList);
    },
    [setFilterBlocks, setSearchValue, blocks],
  );

  return (
    <div>
      <InputField
        label={<FormattedMessage {...messages.search} />}
        name="search"
        onChange={onChangeSearch}
        placeholder={intl.formatMessage(messages.enterBlockName)}
        required
        value={searchValue}
      />
      <ValuePickerProvider name={name} onChange={onChange} value={value}>
        {filterBlocks.map((block) => (
          <GUIEditorBlockItem key={block.name} value={block} />
        ))}
      </ValuePickerProvider>
    </div>
  );
}
