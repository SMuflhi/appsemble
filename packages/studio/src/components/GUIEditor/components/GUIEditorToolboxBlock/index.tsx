import { Icon, Input } from '@appsemble/react-components';
import type { BlockManifest } from '@appsemble/types';
import { stripBlockName } from '@appsemble/utils';
import classNames from 'classnames';
import React from 'react';
import { FormattedMessage, useIntl } from 'react-intl';

import type { NamedEvent } from '../../../../types';
import styles from './index.css';
import messages from './messages';

interface GUIEditorToolboxBlockProps {
  blocks: BlockManifest[];
  name: string;
  onChange: (event: React.ChangeEvent<HTMLInputElement>, block: BlockManifest) => void;
  value: BlockManifest;
}

export default function GUIEditorToolboxBlock({
  blocks,
  name,
  onChange,
  value,
}: GUIEditorToolboxBlockProps): React.ReactElement {
  const [searchValue, setSearchValue] = React.useState<string>('');
  const [filterBlocks, setFilterBlocks] = React.useState<BlockManifest[]>(blocks);

  const intl = useIntl();

  const onChangeSearch = React.useCallback(
    (_event: NamedEvent, query: string) => {
      setSearchValue(query);
      const updatedList = blocks.filter(
        (item) => item.name.toLowerCase().search(query.toLowerCase()) !== -1,
      );

      setFilterBlocks(updatedList);
    },
    [setFilterBlocks, setSearchValue, blocks],
  );

  return (
    <div>
      <Input
        label={<FormattedMessage {...messages.search} />}
        name="search"
        onChange={onChangeSearch}
        placeholder={intl.formatMessage(messages.enterBlockName)}
        required
        value={searchValue}
      />
      {filterBlocks.map((block) => (
        <label
          key={block.name}
          className={classNames('card', styles.blockFrame, {
            [styles.selected]: value === block,
          })}
        >
          <div className="card-content">
            <Icon icon="box" size="medium" />
            <span className="subtitle">{stripBlockName(block.name)}</span>
            <input
              checked={value ? value.name === block.name : false}
              hidden
              name="type"
              onChange={(event) => onChange(event, block)}
              type="radio"
              value={name}
            />
          </div>
        </label>
      ))}
    </div>
  );
}
