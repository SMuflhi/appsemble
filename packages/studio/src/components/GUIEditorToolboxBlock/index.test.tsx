import { Loader } from '@appsemble/react-components';
import { shallow } from 'enzyme';
import React from 'react';

import GUIEditorToolboxBlock, { Block } from '.';

const blocks: Block[] = [
  {
    id: 'map',
    description: 'map description',
    iconName: 'map',
  },
  {
    id: 'navigation',
    description: 'navigation description',
    iconName: 'route',
  },
  {
    id: 'markdown',
    description: 'markdown description',
    iconName: 'font',
  },
];

describe('<GUIEditorToolboxBlock /> rendering', () => {
  let wrapper: any;

  describe('with no properties', () => {
    beforeEach(() => {
      wrapper = shallow(<GUIEditorToolboxBlock />);
    });

    it('should render <Loader /> without given blocks', () => {
      expect(<Loader />).toMatchSnapshot();
    });

    it('should not render a div', () => {
      expect(wrapper.find('div')).toHaveLength(0);
    });
  });

  describe('with blocks values', () => {
    const initialProps = {
      blocks,
      initialSelectedBlock: '',
      selectedBlock(): void {},
    };

    beforeEach(() => {
      wrapper = shallow(<GUIEditorToolboxBlock {...initialProps} />);
    });

    it('should render a div with the main class', () => {
      expect(wrapper.find('.main')).toHaveLength(1);
    });

    it('should render a div with blockframe class for every found block', () => {
      expect(wrapper.find('.blockFrame')).toHaveLength(blocks.length);
    });

    it('should render an icon for every block', () => {
      expect(wrapper.find('Icon')).toHaveLength(blocks.length);
    });

    it('should render the title for every block', () => {
      expect(wrapper.find('h2')).toHaveLength(blocks.length);
    });

    it('should change block class to blockFrameSelected when clicked', () => {
      wrapper.find('.blockFrame').last().simulate('click');
      expect(wrapper.find('.blockFrameSelected')).toHaveLength(1);
      expect(wrapper).toMatchSnapshot();
    });
  });
});
