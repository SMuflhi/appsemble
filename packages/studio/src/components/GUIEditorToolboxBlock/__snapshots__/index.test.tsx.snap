// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GUIEditorToolboxBlock /> rendering with blocks values should change block class to blockFrameSelected when clicked 1`] = `
<div
  className="main"
>
  <div
    className="blockFrame"
    key="map"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <Icon
      icon="map"
      size="large"
    />
    <h2
      className="subtext"
    />
  </div>
  <div
    className="blockFrame"
    key="navigation"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <Icon
      icon="route"
      size="large"
    />
    <h2
      className="subtext"
    />
  </div>
  <div
    className="blockFrameSelected"
    key="markdown"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <Icon
      icon="font"
      size="large"
    />
    <h2
      className="subtext"
    />
  </div>
</div>
`;

exports[`<GUIEditorToolboxBlock /> rendering with no properties should render <Loader /> without given blocks 1`] = `<Loader />`;
